### ISSUE TITLE
Increase playback timeout

### ISSUE THREAD

**Issue by Nazar78** [jellyfin/jellyfin-tizen#153]
Hi,

Is there any way to increase the media operations timeout, play/fwd/rwd etc? I'm asking because it seems only with JF-Tizen I have this issue, but not with the JF-Web.

Here's the deal, JF server is running off an SSD on my NAS which is set to never sleeps on top of others like reverse proxy, Debian 10 OS stuffs etc. I also have web assets and images cached on the RAM so everything is snappy.

The media is however stored on a RAID 10 mechanical 4x10TB 7200RPM 256MB cache disks which often I would want them to go on standby when not in use (reduce power usage and temperature). 

But when I click play on the TV with transcoding or remuxing is required (direct play seems fine), it often ends up in an infinite loading circle when the disks are still asleep awaiting them all to each spin up then resume. When this happened, I can no longer interact with the UI, only way is to force close by holding the TV remotes' back button. Of course when I open the app again, I can resume playback almost immediately as the disks are already awakened.

I just hope there a way to increase this timeout to accommodate the sleeping disks + the time it requires to buffer the transcoding/remuxing.

**Comment by dmitrylyzo** [comment]
Not sure if there is a configurable timeout in HtmlMediaElement. Last time I tested Tizen app with the server not responding, it took 30sec to throw an error. Then it retried different transcoding settings (2-3 times in total).

We don't have a custom timeout, iirc.

How long have you waited?

**Comment by Nazar78** [comment]
Hi @dmitrylyzo,

Thanks for replying.

> How long have you waited?

I waited about 10 secs +/- for all the 4 disks to resume while JF-Tizen is showing the circle animation after pressing play. I knew the disks had resumed when its LEDs starts blinking, also from checking with `hdparm -C /dev/sd[a-d]` showing active/idle instead of standby. However at this point, the GUI is no longer responding so I'm stuck at the circle animation with the media's photoart  background for several minutes and the only way is to force close the app by holding the remote's back button.

> Does the server actually start transcoding? Maybe it just doesn't respond at all.

Yes the transcoding did kicked in, I could see the ffmpeg consuming resources and the transcoding log/tmp files exists. I tried to simulate this by forcing my disks to go on standby with `hdparm -y /dev/sd[a-d]`, ensuring it's really on standby, then click play but unfortunately I couldn't produce the same issue, probably the disk wasn't in deep sleep long enough? I'm not sure it's quite tough to test with external users randomly accessing the server. But I'm very sure the issue happened almost every time I see my disks LEDs pulsing indicating it's on standby.

> Last time I tested Tizen app with the server not responding, it took 30sec to throw an error. Then it retried different transcoding settings (2-3 times in total).

I have never seen this, do you have any idea how to simulate the server not responding to a playback behavior without actually interrupting the connections? Probably then I can use it to debug what's going on. I was thinking of replacing the ffmpeg with an infinite sleep loop (I did this previously using perl to intercept then regex replace the output parameters adding `-noautorotate`, this was when I had video rotating issues long ago with `-hwaccel` which seems to be solved on the recent JF ffmpeg) but that wasn't the case because the transcoding did actually occurred.

Thank you.

**Comment by dmitrylyzo** [comment]
> However at this point, the GUI is no longer responding so I'm stuck at the circle animation with the media's photoart background for several minutes and the only way is to force close the app by holding the remote's back button.

Looks like the TV doesn't retry and just waits for a response.
If there is no error in the browser console, we may need a custom timeout.
Maybe we can manually make the player to retry several times until an error occurs or playback starts.

> probably the disk wasn't in deep sleep long enough?

Maybe a disk cache in RAM.
No need to stress HDDs by start/stop. This can probably be reproduced by placing a breakpoint when debugging the server, but you need to setup some C# IDE (VS Code) + install .NET 6.

> I have never seen this

There was a bug - https://github.com/jellyfin/jellyfin/pull/6558.

> from https://github.com/jellyfin/jellyfin/pull/6558:
> Ideally, we should probably add a client-side timeout as well.

Now my comment makes more sense.

> do you have any idea how to simulate the server not responding to a playback behavior without actually interrupting the connections?

When Remuxing, DirectStreaming or Transcoding, the TV requests:
1. `master.m3u8` - contains a link to the actual stream (`main.m3u8`).
2. `main.m3u8` - list of TS (or fMP4) segments.
3. `*.ts` or `*.mp4` segments.

You can try to catch them with Wireshark. I don't remember if these requests are visible in the Network tab.
Maybe you can stall the request with a reverse proxy. :thinking:

**Comment by Nazar78** [comment]
> Maybe a disk cache in RAM.
Make sense, thank you, slipped my mind about flushing this out.

Those files were definitely created but I haven't checked if they made it out to the TV.

Thanks again for the tips, I'll check them during off peak period.

**Comment by Nazar78** [comment]
Hi again @dmitrylyzo,

With the recent version 10.8.9, it seems to be getting worst. Playing any media even with Direct Play will result in termination during playback. I noticed this every time when the disks went to sleep, the playback terminates back to the previous page whichever it was before starting to play. The tracktime goes back to zero as seems the media hasn't been played.

This issue doesn't seems to happen on the Chrome browser or the Android client. I also tried in the emulator 5.0 then forcing the disks to sleep `hdparm -y /dev/sd[a-d]` but I couldn't replicate it similar to the physical TV. Tried looking in the JF server logs, nothing seems significant.

This is really strange. I always get interrupted when playing on the TV, then need to replay scrub back to where I roughly stopped. I've tried both reverse-proxy and direct LAN IP no matter it still happened on the TV. Do you have any more other advise?

I do have two other options which I'm still contemplating:
1. Increase my disks timeout, currently 30mins.
2. Roll back to the previous JF server version.

**Comment by dmitrylyzo** [comment]
Wait, does the HDD go into sleep while watching? 👀

If it is not DirectPlay, I would say that something stops the transcoding and the playback position becomes equal to the duration, which means "watched".

I was able to reproduce your original problem with the UI not responding. I put a breakpoint in the server when debugging, and after continuing, the app is stuck.
So we definitely need a timeout (in web), which will restart the playback (probably several times) if it doesn't start on time.

**Comment by Nazar78** [comment]
> Wait, does the HDD go into sleep while watching? 👀

Yes, strange but I'm afraid it does. I actually have two separate locations on my Asustor AS55304T NAS, JF server (nginx proxied) running on a single external USB SSD which never goes to sleep as there's many other apps running on the SSD like docker, LXC, websites etc. This SSD is intended not to go to sleep. Then all the media resides on an internal RAID10 4xHDDs which is set to sleep after 30mins. There's no issue at all if I'm playing using other players such as VLC on Ubuntu Desktop running on the same SSD accessing media on the same HDDs. So there's definitely something wrong with JF Tizen or my TV. I did went check the change log, nothing significant that could cause this. Thinking of it, tomorrow I'll probably test and install Plex (was using this long ago) or Emby see if it happens too 🙄

> If it is not DirectPlay, I would say that something stops the transcoding and the playback position becomes equal to the duration, which means "watched".

Yes you're right, it's watched. Now with the latest version it doesn't matter Direct Play or not, during playback the screen will freeze momentarily then pops back to the previous page and I lost track of the timeline.

> So we definitely need a timeout (in web), which will restart the playback (probably several times) if it doesn't start on time.

Thank you for this, really hope you can help do something about it. As for now I'll just use VLC.

**Comment by Nazar78** [comment]
![image](https://user-images.githubusercontent.com/2883688/216463192-a6d6066f-b665-4329-bb0b-7cbcf3d8aae7.png)

The emulator Tizen 5.0 still kept going even the disks went to sleep several times. With the same Direct Play on the TV I always get the same issue as described.

**Comment by Nazar78** [comment]
Hi @dmitrylyzo,

I installed and tested Emby on my NAS. I installed the server binaries and config/metadata running on the external USB SSD.  Tried doing a DirectPlay for a media that's from the HDDs. Unfortunately I can't seem to find any Emby Tizen Client for my TV in the app store so I tried using JF Client instead and behold, no issues for the whole 2hrs+ of media playback.

Do note that I just previously tested the same DirectPlay media on the JF server 10.8.9 and roughly when my disks went to sleep in  about 30mins, it froze momentarily then back to the previous page and marked as watched when I revisit the page again.

So this is not a client (despite not doing any auto retries) but a server issue on the latest version? There's nothing obvious in the JF server logs that I could report.

**Comment by dmitrylyzo** [comment]
I don't think the app can do anything with the server's HDDs. Only the server has direct access and wakes them up. And it is weird that they go sleep, even being accessed. Can't it be due to nginx caching? Or you don't cache video?

I suggest you build Tizen app with `release-10.8.x` branch. This way it will be more compatible with the server and behave more like an Android app or in a browser (they get the UI from the server).

**Comment by G-eos** [comment]
I have some trouble when I play video which are stored remotely from a far away host. This is a huge ugly hack, based on curlftpfs mounted as a docker volume :-) ... sometimes, the video stop and then restart and then stop 2 min later, pause 10 seconds ... and other times it play a 4k video smoothly for hours.

I never find the exact condition to trig the issue. Network latency and bandwidth is not the reason. Since latest JF version, the occurrence reduced.

I have no idea if this is the same issue but I always think that it may be related to buffering which are not enough. Sorry no logs to share, when I had look on it I do not find clear items showing things. ... 

If it helps, I can retry and get logs, please define which one.

**Comment by Nazar78** [comment]
> I don't think the app can do anything with the server's HDDs. Only the server has direct access and wakes them up. And it is weird that they go sleep, even being accessed. Can't it be due to nginx caching? Or you don't cache video?

I told Nginx only to cache the web assets and images. They're cached in tmpfs so caching video will quickly fill up my maxed limited 16GB RAM which I wouldn't want it to happen. The issue also happened in direct LAN IP without going through Nginx.

Since it got worst in the latest JF 10.8.9, I would suspect buffering issues, somehow buffer was increased so less often the media is being read from the disk, but again nothing similar mentioned in the change logs.

> I suggest you build Tizen app with `release-10.8.x` branch. This way it will be more compatible with the server and behave more like an Android app or in a browser (they get the UI from the server).

I did tried that for the previous issue and thank you for reminding I'll try again. Worst case I'll just increase my disk sleep timeout.

**Comment by Nazar78** [comment]
@G-eos 
> I have no idea if this is the same issue but I always think that it may be related to buffering which are not enough.

My media is on the same premise and device but different disk mount, server on USB SSD and media on internal HDDs. Me on the other hand thinks that the buffer is too much that it timeout my HDDs to sleep. Not really sure though just an assumption. I agree with @dmitrylyzo that not much either the app (nor server) can do however some client auto retry reconnect attempts in the background would be helpful for our case.

**Comment by Nazar78** [comment]
Hi @dmitrylyzo,

I've tried the Tizen app build `release-10.8.x` I still see the issue. This is a direct play no ffmpeg log created.

Here's the JF server debug log leading to the issue, I split them into two parts, start and end, because nothing significant in between:

[start.txt](https://github.com/jellyfin/jellyfin-tizen/files/10610099/start.txt)

[end.txt](https://github.com/jellyfin/jellyfin-tizen/files/10610100/end.txt)

The part "Playback stopped reported by app" is where the playback was playing as per normal then suddenly stopped (without me doing anything even to the remote) and returned to the previous page.

**Comment by Nazar78** [comment]
An update, I was all over the place trying to figure out what's going on and found the culprit. Apparently it was my Samsung RU8000 5GHz Wi-Fi that's causing the issue. Not sure if it's my RT-AX86U Merlin firmware router or the TV itself but I tested tethering using my phone using 5GHz and it does the same thing so probably something is wrong with the TV. Switched to 2.4GHz it went away. I have another new spare RT-AX86U or some few older routers which I'll test later when time permits.

Still I hope the "retry" can be introduced because the same situation happened (got kicked back to previous page) when I tested debugging in Chrome setting the network to offline while playing.
